### DISCUSSION TITLE
How to control a shutter with a three-button command

### DISCUSSION THREAD

**Discussion by pin010** [arendst/Tasmota#11496] [Q&A] [answered]
Hi, I'm new to GitHub, so far I have dabbled with PICxxx designs and programming. A few months ago I discovered home automation, Tasmota and ESP8266.
I have a remote controlled awning, which is it.
Somfy Situo 5 IO pure
 freq. 868.95 MHz
![smfytc](https://user-images.githubusercontent.com/81494843/112747426-a9780600-8fb5-11eb-8b35-e4ff13c813fe.jpeg)
All buttons work by pressing them for a moment without having to keep them pressed
by pressing the "up" button, the curtains retract completely
by pressing the "my" button, the movement of the curtains stops
by pressing the "down" button, the curtains extend completely

I intercepted the points to connect in parallel to the buttons inside the case
![P2260100a](https://user-images.githubusercontent.com/81494843/112747454-e04e1c00-8fb5-11eb-8ed7-57452d4d4e90.jpg)

and I connected four conductors to make this schematic
![opto](https://user-images.githubusercontent.com/81494843/112747474-1b504f80-8fb6-11eb-935e-19964fa6e1c2.jpg)
GPIO5 -> D1 -> UP
GPIO4   -> D2 -> DOWN
GPIO12 -> D6 -> STOP

I have configured my esp8266 like this

![tmplt](https://user-images.githubusercontent.com/81494843/112747678-33749e80-8fb7-11eb-83c6-a8cf95cc479c.png)

{"NAME":"Tende","GPIO":[1,1,1,1,224,225,1,1,226,1,1,1,1,1],"FLAG":0,"BASE":30}
PulseTime1 1..2
PulseTime2 1..2
PulseTime3 1..2

![tende](https://user-images.githubusercontent.com/81494843/112747758-a8e06f00-8fb7-11eb-91c3-418497794301.png)

I can then through MQTT to connect everything with OpenHAB and wanting to Alexa.
Everything works, but I would like to use it with the calibration as shown in

![mygoal](https://user-images.githubusercontent.com/81494843/112747222-2c985c80-8fb4-11eb-990d-d239755defa9.png)
Is there any way to adapt my device to this way of working?
Thank you very much!

**Comment by TrDA-hab** [comment, marked as answer]
1. No, this is not possible - you do not control the motor, you give control impulses for STM8 (PulseTime1 1..2). The motor is controlled by STM8.
2. Control via "ShutterMode" is possible only through a driver connected directly to the motor (Stepper motor, DC motor, Servo motor).
3. You can read how to directly control the DC motor here:
https://github.com/arendst/Tasmota/discussions/10387

**Reply by pin010** [reply]
Thank you for your answer :-)
I'm afraid I have explained myself badly.
STM8 controls the 868.95Mhz transmitter of the remote control, I imagine that the receiver near the motor, perhaps at 220V ac, uses two self-holding relays and that the stop command releases them. (a bit like it happened with motors controlled by contactors, in the distant Pliocene)

**Reply by TrDA-hab** [reply]
Now I understand you, you have repeated this project:
https://www.instructables.com/Somfy-Control-From-Your-Mobile-IFTTT-and-Google-fo/
The motor is directly controlled by another controller via a relay. If you connect the ESP directly to the relay inputs, you can control the motor via "ShutterMode".

**Reply by pin010** [reply]
Thanks, I did not know instructables.com
and yes, I did the same job, I have the three buttons on openHAB and also on Alexa, but the voice command becomes: "Alexa [turn on] 'raise awnings'" while it would be better "Alexa [raise]' awnings '", another problem is that there is no command" raise the curtains by 50% "because we are facing three switches not the shutter command. Not very elegant.
I cannot access the motor, it is in a dangerous position and the 230V power supply comes directly to the junction box of the awning, there are no wires or buttons on the wall.
So I'm thinking of interposing between ESP (programmed "ShutterMode") and photocoupler, a PIC12F1822 or similar, which will have to produce the following timing:
![TIMING](https://user-images.githubusercontent.com/81494843/112767925-a9115680-9019-11eb-9e66-7b0de55693e2.png)

(I have a pseudoC demo code snippet ready but I don't know how to post it without losing formatting)

**Reply by TrDA-hab** [reply]
Good luck !!!

**Reply by lwyl5744** [reply]
I would like more information on the v2.0 Solar and Weather monitoring system if this is possible

**Comment by pin010** [comment]
Second round, I set up my ESP exactly as described in the great article [HOWTO control DC motors](https://github.com/arendst/Tasmota/discussions/10387) 1. MX1508 Motor Driver.
As a proof of concept, I made the previously proposed scheme, flashed the only PIC I had at home with the program below and in the breadboard checked that the remote control buttons received the correct impulses.
```C
void main(void)
{
    // initialize the device
    SYSTEM_Initialize();
    char trigUP = 0;
    char trigDOWN = 0;
    while(1)
    {
        if ((IO_inUP_GetValue()==1)&&(IO_inDOWN_GetValue()==1)) // Something went wrong
        {
            IO_outUP_SetLow(); //outUP=0;
            IO_outDOWN_SetLow(); //outDOWN=0;
            IO_outSTOP_SetLow(); //outSTOP=0;
        }
        else
        {
            if ((IO_inUP_GetValue()==1)&&(trigUP==0)&&(trigDOWN==0))
            {
                IO_outUP_SetHigh();	// started UP movement
                __delay_ms(200); //200 ms is the time we keep the button pressed
                IO_outUP_SetLow();	// this don't stop movement
                trigUP=1;
            }
            if ((IO_inDOWN_GetValue()==1)&&(trigDOWN==0)&&(trigUP==0))
            {
                IO_outDOWN_SetHigh();
                __delay_ms(200); //200 ms is the time we keep the button pressed
                IO_outDOWN_SetLow();
                trigDOWN=1;
            }
        }
        if ((trigUP==1)&&(IO_inUP_GetValue()==0)) // stopped keeping UP active
        {
            IO_outSTOP_SetHigh();	// so I stop the movement
            __delay_ms(200); // 200 ms is the time we keep the button pressed
            IO_outSTOP_SetLow();
            trigUP=0;
        }
        if ((trigDOWN==1)&&(IO_inDOWN_GetValue()==0)) // stopped keeping DOWN active
        {
            IO_outSTOP_SetHigh();	// so I stop the movement
            __delay_ms(200); // 200 ms is the time we keep the button pressed
            IO_outSTOP_SetLow();
            trigDOWN=0;
        }
    }
}	

```
Now it seems redundant to use another microcontroller to do what the ESP8266 could surely do, but I am not able to intervene in the Tasmota code.
Would anyone help me?

**Comment by pin010** [comment]
Nobody?
Then it is useless to go beyond the good proof obtained in the breadboard, 

![IMG_3610](https://user-images.githubusercontent.com/81494843/113692972-6234f800-96ce-11eb-9a29-3f92e3dcd4de.JPEG)

my device remains like this, 
![IMG_3594](https://user-images.githubusercontent.com/81494843/113692478-dfac3880-96cd-11eb-9bc2-2c61e4e9dd7f.JPEG)
![IMG_3593](https://user-images.githubusercontent.com/81494843/113694280-d3c17600-96cf-11eb-8883-1e96d9e64c2c.JPEG)
![IMG_3591](https://user-images.githubusercontent.com/81494843/113694208-c0aea600-96cf-11eb-9207-e9063e65a407.JPEG)


it will flash differently. Thanks.

**Comment by pin010** [comment]
Hi, it turned out that I used the simplest and most immediate system possible, I used the HUE emulation and no HUB, Alexa connects directly and finds three devices, STOP, DOWN and UP.
Then you can create your own routine.
You need an ECHO device, my mobile app alone is not enough. So I had to go and finish the job at my son's house ... after all, the curtains belong to him.
